src/main/java/org/jitsi/videobridge/cc/vp8/VP8FrameMap.java

```diff
@@ -0,0 +1,259 @@
+
+import java.util.*;
+import java.util.concurrent.*;
+
+/**
+ * A history of recent frames on a VP8 stream.
+ */
+public class VP8FrameMap
+{
+    private final ConcurrentSkipListMap<Integer, VP8Frame>
+        vp8FrameMap = new ConcurrentSkipListMap<>(
+        /* This is only a valid Comparator if seq number diffs of all
+         * timestamps are within half the number space.  (This property is
+         * assured by #cleanupFrameMap.)
+         */
+        (s1, s2) -> (int) RtpUtils.getSequenceNumberDelta(s1, s2));
+
+    private final Logger logger;
+
+    /**
+     * The diagnostic context of this instance.
+     */
+    private final DiagnosticContext diagnosticContext;
+
+
+    private final int FRAME_MAP_SIZE = 500; /* Matches PacketCache default size. */
+
+    /**
+     * Ctor.
+     *
+     */
+    public VP8FrameMap(
+        @NotNull DiagnosticContext diagnosticContext,
+        @NotNull Logger parentLogger)
+    {
+        this.diagnosticContext = diagnosticContext;
+        this.logger = parentLogger.createChildLogger(VP8AdaptiveTrackProjectionContext.class.getName());
+    }
+
+    /** Clean up old entries in the frame map.
+     *
+     * The frame map is maintained as a sorted map; to clean it up,
+     * we walk entries from the oldest (in sequence number ordering), removing anything
+     * more than #FRAME_MAP_SIZE older than the newest seq.
+     *
+     * Because the map is ordered, we can stop searching entries when we reach
+     * one that is new enough not to be removed, so this is amortized constant
+     * time per frame.
+     *
+     * Caller should be synchronized.
+     *
+     * Return false if this sequence is too old.
+     * */
+    private boolean cleanupFrameMap(int newSeq)
+    {
+        if (vp8FrameMap.isEmpty())
+        {
+            return true;
+        }
+
+        int latestSeq = vp8FrameMap.lastKey();
+
+        /* If our sequence numbers have jumped by a quarter of the number space, reset the map. */
+        int seqJump = RtpUtils.getSequenceNumberDelta(newSeq, latestSeq);
+        if (seqJump >= 0x4000 || seqJump <= -0x4000)
+        {
+            vp8FrameMap.clear();
+            return true;
+        }
+
+        int threshold = RtpUtils.applySequenceNumberDelta(latestSeq, -FRAME_MAP_SIZE);
+
+        if (RtpUtils.isOlderSequenceNumberThan(newSeq, threshold))
+        {
+            return false;
+        }
+
+        Iterator<Integer> it = vp8FrameMap.keySet().iterator();
+        while (it.hasNext())
+        {
+            Integer key = it.next();
+            if (RtpUtils.isOlderSequenceNumberThan(key, threshold))
+            {
+                it.remove();
+            }
+            else {
+                break;
+            }
+        }
+
+        return true;
+    }
+
+
+    /** Find a frame in the frame map, based on a packet. */
+    public synchronized VP8Frame findFrame(Vp8Packet packet)
+    {
+        int seq = packet.getSequenceNumber();
```

Comment:
why does this use sequence number and not timestamp?

Reply:
To make "previous frame" and "next frame" well-defined.  Technically, timestamps aren't required to be monotonic with respect to sequence numbers (though this is a feature that isn't often used for interactive video, because it adds encoding delay).

Reply:
ah ok, i just thought it was the timestamp which separated/distinguished frames

Reply:
Yeah, it's the timestamp that tells you whether two packets are the same frame, but it's the sequence numbers that tell you the decode order.

Reply:
This function does two things -- it finds the frame to which a packet belongs, if it's already recorded; or if it's not it creates a new frame and finds the previous and next frames.

Probably should be documented/named better.

Reply:
Following up on this - you'll see later in the same file there are now a bunch of functions to get the previous and next frames, either absolutely or with various condition tests.

I didn't see any reason to maintain two data structures (one mapping packets to frames, another maintaining frames in decode order) when I could do the job with one.